### PR TITLE
Added functionality in convertCollectionToIDMap that throws the IdShortDuplicationException

### PR DESCRIPTION
**Pull request by mdanish98**
Dear Frank

I have added functionality in SubmodelElementMapCollectionConverter.convertCollectionToIDMap that throws the IdShortDuplicationException if there are two elements with the same IdShort.

Thanks and Regards
Mohammad Ghazanfar Ali Danish

### CODE REVIEW

**Comment by FrankSchnicke** [comment]
Dear Mohammad,

many thanks. However, your MQTT changes are somehow included. Can you fetch upstream and merge with your working state? Afterwards, I will add my review.

Best regards,
Frank

**Comment by mdanish98** [comment]
Dear Frank,

Thank you for the input.
I did fetch upstream and merged with my working state.

Thanks and Regards
Mohammad

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java [hunk 1]
```diff
@@ -85,4 +87,27 @@ public void testCreateAsFacadePropertyList() {
...
+	
+	@Test(expected = IdShortDuplicationException.class)
+	public void checkForExceptionWithDuplicateIdShortInSubmodel() {
+		Map<String, Object> faultySubmodel = createSubmodelWithDuplicateIdShortProperties();
+
+		Submodel.createAsFacade(faultySubmodel);
+	}
+
+	private Map<String, Object> createSubmodelWithDuplicateIdShortProperties() {
+		Property property1 = new Property("testProp", 5);
```
Comment (FrankSchnicke) [inline src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java:99]: Please extract the common IdShort in a String variable called, e.g., duplicateIdShort

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java [hunk 2]
```diff
@@ -85,4 +87,27 @@ public void testCreateAsFacadePropertyList() {
...
+		Property property2 = new Property("testProp", 7);
+		
+		Collection<Map<String, Object>> collection = new ArrayList<>();
```
Comment (FrankSchnicke) [inline src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java:102]: You can use Arrays.asList(property1, property2) here

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java [hunk 3]
```diff
@@ -85,4 +87,27 @@ public void testCreateAsFacadePropertyList() {
...
+		
+		collection.add(property1);
+		collection.add(property2);
+		
+		Submodel submodel = new Submodel();
```
Comment (FrankSchnicke) [inline src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java:107]: Please add an arbitrary idShort & identifier to ensure constructing a valid submodel

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java [hunk 4]
```diff
@@ -212,6 +212,11 @@ public void testDeleteSubmodelElementNotExist() {
 		collection.deleteSubmodelElement("Id_Which_Does_Not_Exist");
 	}
 	
+	@Test
```
Comment (FrankSchnicke) [inline src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java:215]: Please remove this empty test

File: src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/SubmodelElementCollection.java [hunk 5]
```diff
@@ -108,6 +108,8 @@ public static SubmodelElementCollection createAsFacade(Map<String, Object> obj)
 		}
 		
 		if (!isValid(obj)) {
+			System.out.println("Exception : obj = " + obj);
```
Comment (FrankSchnicke) [inline src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/SubmodelElementCollection.java:111]: Please remove these two prints to command line

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java [hunk 6]
```diff
@@ -263,4 +268,28 @@ private Operation getOperation() {
...
+	
+	@Test(expected = IdShortDuplicationException.class)
+	public void checkForExceptionWithDuplicateIdShortInSubmodelElementCollection() {
+		Map<String, Object> faultySubmodelElementCollection = createSubmodelElementCollectionWithDuplicateIdShortProperties();
+
+		SubmodelElementCollection.createAsFacade(faultySubmodelElementCollection);
+	}
+
+	private Map<String, Object> createSubmodelElementCollectionWithDuplicateIdShortProperties() {
+		Property property1 = new Property("testProp", 5);
```
Comment (FrankSchnicke) [inline src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java:280]: See comments testcase above

File: src/main/java/org/eclipse/basyx/aas/metamodel/exception/IdShortDuplicationException.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ ******************************************************************************/
+package org.eclipse.basyx.aas.metamodel.exception;
+
+import java.util.Map;
+
+/**
+ * This class is used to throw exception when there are two
```
Comment (FrankSchnicke) [inline src/main/java/org/eclipse/basyx/aas/metamodel/exception/IdShortDuplicationException.java:15]: This class is not used to throw an exception - it _represents_ this exceptions

File: src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java [hunk 8]
```diff
@@ -151,9 +152,18 @@ public static SubmodelElementCollection mapToSmECollection(Map<String, Object> s
 			for (Object o: smElementsSet) {
 				Map<String, Object> smElement = (Map<String, Object>) o;
 				String id = (String) smElement.get(Referable.IDSHORT);
+				
+				if(smElementsMap.containsKey(id)) {
+					throw new IdShortDuplicationException(SubmodelElementMapCollectionConverter.class, (Map<String, Object>) smElementsMap);
```
Comment (FrankSchnicke) [inline src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java:157]: The class passed to the constructor is the class that was intended to construct. Thus, passing SubmodelElementMapCollectionConverter.class does note make any sense. At the same time, we don't know which kind of element should be created here. In consequence, please add a constructor in MetamodelCreationException that takes an error message and add one constructor in IdShortDuplicationException that takes a map and calls this super constructor with a message like "Could not construct meta model element due to IdShort duplication. Passed argument was " + map.toString());

File: src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java [hunk 9]
```diff
@@ -166,6 +176,30 @@ public static SubmodelElementCollection mapToSmECollection(Map<String, Object> s
...
 	}
 	
+	/**
+	 * Checks if there are two elements in the provided Submodel Elements with the same IdShort <br>
+	 * 
+	 * @param Submodel Elements
+	 * @return True if there is duplicate IdShort present, otherwise False.
+	 */
+	@SuppressWarnings("unchecked")
+	private static boolean isDuplicateIdShortPresentInSubmodel(Object submodelElements) {
```
Comment (FrankSchnicke) [inline src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java:186]: This method should be named isDuplicateIdShortPresentInSubmodelElements. Additionally, please use a Map<String, Object> as input parameter and perform the cast on the caller's side

**Review by FrankSchnicke — CHANGES_REQUESTED** [review]

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java (hunk 2 above)
Comment (mdanish98) [reply src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java]: Updated the code.

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java (hunk 3 above)
Comment (mdanish98) [reply src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/TestSubmodel.java]: Added an idShort & identifier.
Updated the code.

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java (hunk 4 above)
Comment (mdanish98) [reply src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java]: Removed this empty test.

File: src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/SubmodelElementCollection.java (hunk 5 above)
Comment (mdanish98) [reply src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/SubmodelElementCollection.java]: Removed this two print lines.

File: src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java (hunk 6 above)
Comment (mdanish98) [reply src/test/java/org/eclipse/basyx/testsuite/regression/submodel/metamodel/map/submodelelement/TestSubmodelElementCollection.java]: Updated the code.

File: src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java (hunk 8 above)
Comment (mdanish98) [reply src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java]: - Added a constructor in MetamodelCreationException that takes an error message
- Added a constructor in IdShortDuplicationException that takes a map

File: src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java (hunk 9 above)
Comment (mdanish98) [reply src/main/java/org/eclipse/basyx/submodel/metamodel/facade/SubmodelElementMapCollectionConverter.java]: - Rename it to isDuplicateIdShortPresentInSubmodelElements
- Performed the cast on the caller's side of isDuplicateIdShortPresentInSubmodelElements

**Review by FrankSchnicke — APPROVED** [review]

**Pull request merged**